framework/src/media/audio/audio_manager.c

```diff
@@ -934,93 +854,117 @@ audio_manager_result_t reset_audio_stream_out(void)
+{
+	struct audio_caps_desc_s caps_desc;
+
+	caps_desc.caps.ac_len = sizeof(struct audio_caps_s);
+	caps_desc.caps.ac_type = AUDIO_TYPE_OUTPUT;
+	if (card_type == INPUT) {
+		caps_desc.caps.ac_type = AUDIO_TYPE_INPUT;
 	}
+	caps_desc.caps.ac_subtype = AUDIO_TYPE_QUERY;
 
-	if (type == INPUT) {
-		g_audio_in_cards[g_actual_audio_in_card_id].resample.samprate_types = caps->ac_controls.b[0];
-		medvdbg("Supported Input sample rate: %d, %x\n", caps->ac_controls.b[0], caps->ac_controls.b[0]);
-	} else {
-		g_audio_out_cards[g_actual_audio_out_card_id].resample.samprate_types = caps->ac_controls.b[0];
-		medvdbg("Supported Output sample rate: %d, %x\n", caps->ac_controls.b[0], caps->ac_controls.b[0]);
+	if (ioctl(fd, AUDIOIOC_GETCAPS, (unsigned long *)&caps_desc.caps) < 0) {
+		meddbg("Fail to ioctl AUDIOIOC_GETCAPS\n");
+		return AUDIO_MANAGER_DEVICE_FAIL;
 	}
 
+	*sample_type = caps_desc.caps.ac_controls.b[0];
+
 	return AUDIO_MANAGER_SUCCESS;
 }
 
-static uint32_t get_closest_samprate(uint32_t origin_samprate, audio_device_type_t type)
+static audio_manager_result_t get_audio_volume(int fd, audio_config_t *config, audio_card_type_t card_type)
 {
-	int i;
-	uint32_t result;
-	uint32_t samprate_types;
-	int count = sizeof(g_audio_samprate_entry) / sizeof(struct audio_samprate_map_entry_s);
+	struct audio_caps_desc_s caps_desc;
+	int max_volume;
+	int cur_volume;
 
-	ASSERT(count > 0);
+	caps_desc.caps.ac_len = sizeof(struct audio_caps_s);
+	caps_desc.caps.ac_type = AUDIO_TYPE_FEATURE;
+	caps_desc.caps.ac_format.hw = AUDIO_FU_VOLUME;
 
-	if (type == INPUT) {
-		ASSERT(g_actual_audio_in_card_id >= 0);
-		samprate_types = g_audio_in_cards[g_actual_audio_in_card_id].resample.samprate_types;
-	} else {
-		ASSERT(g_actual_audio_out_card_id >= 0);
-		samprate_types = g_audio_out_cards[g_actual_audio_out_card_id].resample.samprate_types;
+	if (card_type == INPUT) {
+		caps_desc.caps.ac_format.hw = AUDIO_FU_INP_GAIN;
 	}
 
-	result = g_audio_samprate_entry[count - 1].samprate;
+	if (ioctl(fd, AUDIOIOC_GETCAPS, (unsigned long *)&caps_desc.caps) < 0) {
+		meddbg("Fail to ioctl AUDIOIOC_GETCAPS\n");
+		return AUDIO_MANAGER_DEVICE_FAIL;
+	}
 
-	for (i = count - 1; i >= 0; i--) {
-		if (g_audio_samprate_entry[i].samprate_types & samprate_types) {
-			if (g_audio_samprate_entry[i].samprate >= origin_samprate) {
-				result = g_audio_samprate_entry[i].samprate;
-			}
-		}
+	max_volume = caps_desc.caps.ac_controls.b[0];
+	cur_volume = caps_desc.caps.ac_controls.b[1];
+
+	/* scale here */
+	cur_volume = cur_volume * AUDIO_DEVICE_MAX_VOLUME / (max_volume - (max_volume % AUDIO_DEVICE_MAX_VOLUME));
+
+	config->max_volume = max_volume;
+	config->volume = cur_volume;
+
+	return AUDIO_MANAGER_SUCCESS;
+}
+
+static void get_hardware_params(audio_card_info_t *card, audio_card_type_t card_type)
+{
+	int fd;
+	int ret;
+	audio_config_t *config;
+
+	fd = open(card->card_path, O_RDONLY);
```

Comment:
"card->card_path" needs to be opened with "O_WRONLY" for output cards.

Reply:
I think you confusing file permission and pcm operation.
Virtual File system consist of flash file system and various drivers.
Each drivers also follow file operation, open, read, write.
If flag is O_RDONLY, file system restrict write operation, for driver case, put some values into driver's priv data to change driver.
If flag is O_WRONLY, file system restrict read operation. This is used when user want get data from driver.
Ioctl doesn't call write function so O_RDONLY is correct.
Also please check this getter, not setter I mean even if we should use read or write, read is correct operation for getter